EIPS/eip-7701.md

```diff
@@ -0,0 +1,213 @@
+
+RIP-7560 is build on the current Ethereum contract structure, and therefore has little choice but to rely
+on using some higher-level abstraction.
+
+In its current form, RIP-7560 transactions use Solidity method selectors in order to achieve this separation.
+
+This, however, is far from ideal as this approach "leaks" the concept from a programming language widely used
+in the EVM into the core design of the Ethereum protocol.
+
+While there is no purely technical reason to disallow it and there are already instances of Solidity code
+being "enshrined" in the Ethereum protocol, e.g. the validator deposit contract, such violation of abstraction levels
+often lead to unnecessary technical debt and are worth avoiding if possible.
+
+Additionally, method selectors provide very weak indication of the contract developer's decision to become a
+participant in Native Account Abstraction transaction flow.
+The chance of accidentally exposing a function with a colliding 4 byte
+method identifier that returns a valid 4 byte magic to indicate approval is pretty low, but a malicious developer
+can easily hide such a function giving it an innocent name, making it hard to spot a Native Account Abstraction entity.
+
+This issue to some extent is also present in [ERC-4337](./eip-4337.md).
+
+As an alternative, if Native Account Abstraction is to be implemented in coordination with [EIP-3540](./eip-3540),
+relying on the concept of "code sections" it introduces is a superior approach.
+
+## Specification
+
+### System-level code entry points
+
+Modify the EOF container format to consist of the following sections:
+
+```
+container := header, body
+header :=
+    magic, version,
+    kind_types, types_size,
+    kind_entrypoints, entrypoints_size,
+    kind_code, num_code_sections, code_size+,
+    [kind_container, num_container_sections, container_size+,]
+    kind_data, data_size,
+    terminator
+body := types_section, entrypoints_section, code_section+, container_section*, data_section
+types_section := (inputs, outputs, max_stack_height)+
+entrypoints_section := (entrypoints_role, target_section_index, target_section_pc_offset)+
+```
+
+For regular calls to the contract, the execution always starts at the first byte of code section 0, and pc is set to 0.
+
+Here the `entrypoints_section` defines alternative indexes of code sections to start the execution for system calls.
+This is reserved for execution of special roles in the `entrypoints_role` range.
+
+Note: do not confuse code execution `entrypoint` with the `EntryPoint` contract defined in ERC-4337.
+
+### Validation and PostTransaction code entry points
+
+The contract that may have a role in an Account Abstraction transaction, either as a Sender, a Paymaster or a Deployer,
+has to contain a section marked with one of the following `entrypoints_role` marker:
+
+```
+role_sender_execution = 0x0000
+role_sender_deployment = 0x0001
```

Comment:
`role_sender...` are exposed by the sender.
This one is `role_deployer_create_sender`

But actually, it is not strictly required: a deployer is just a contract, called with some callData. It doesn't have any restrictions on its execution, so it doesn't require a special eof role